Marlin/src/core/types.h

```diff
@@ -0,0 +1,398 @@
+  bool      operator!=(const XYval<T>   &xy)         { return !operator==(xy); }
+  bool      operator!=(const XYZval<T>  &xyz)        { return !operator==(xyz); }
+  bool      operator!=(const XYZEval<T> &xyze)       { return !operator==(xyze); }
+  bool      operator!=(const XYval<T>   &xy)   const { return !operator==(xy); }
+  bool      operator!=(const XYZval<T>  &xyz)  const { return !operator==(xyz); }
+  bool      operator!=(const XYZEval<T> &xyze) const { return !operator==(xyze); }
+  XYval<T>       operator-()                         { XYval<T> o = *this; o.x = -x; o.y = -y; return o; }
+  const XYval<T> operator-()                   const { XYval<T> o = *this; o.x = -x; o.y = -y; return o; }
+};
+
+template<typename T>
+struct XYZval {
+  union {
+    struct { T x, y, z; };
+    struct { T a, b, c; };
+    T pos[3];
+  };
+  void set(const T px)                                { x = px; }
+  void set(const T px, const T py)                    { x = px; y = py; }
+  void set(const T px, const T py, const T pz)        { x = px; y = py; z = pz; }
+  void set(const XYval<T> pxy, const T pz)            { x = pxy.x; y = pxy.y; z = pz; }
+  void reset()                                        { x = y = z = 0; }
+  T length()                                    const { return (T)sqrtf(x*x + y*y + z*z); }
+  operator T* ()                                      { return pos; }
+  operator bool()                                     { return z || x || y; }
+  XYZval<T>          copy()                     const { XYZval<T> o = *this; return o; }
+  XYZval<int16_t>   asInt()                           { XYZval<int16_t> o = { int16_t(x), int16_t(y), int16_t(z) }; return o; }
+  XYZval<int16_t>   asInt()                     const { XYZval<int16_t> o = { int16_t(x), int16_t(y), int16_t(z) }; return o; }
+  XYZval<int32_t>  asLong()                           { XYZval<int32_t> o = { int32_t(x), int32_t(y), int32_t(z) }; return o; }
+  XYZval<int32_t>  asLong()                     const { XYZval<int32_t> o = { int32_t(x), int32_t(y), int32_t(z) }; return o; }
+  XYZval<float>   asFloat()                           { XYZval<float>   o = {   float(x),   float(y),   float(z) }; return o; }
+  XYZval<float>   asFloat()                     const { XYZval<float>   o = {   float(x),   float(y),   float(z) }; return o; }
+  XYZval<float> reciprocal()                    const { XYZval<float>   o = {  _RECIP(x),  _RECIP(y),  _RECIP(z) }; return o; }
+  XYZval<float> asLogical()                     const { XYZval<float> o = asFloat(); toLogical(o); return o; }
+  XYZval<float>  asNative()                     const { XYZval<float> o = asFloat(); toNative(o); return o; }
+  XYZval<T> ABS()                               const { XYZval<T> o; o.set(_ABS(x), _ABS(y), _ABS(z)); return o; }
+  operator XYval<T>&()                                { return *(XYval<T>*)this; }
+  operator XYval<T>&()                          const { return *(XYval<T>*)this; }
```

Comment:
This allows implicit conversion of a const XYZval to non-const XYval. Similar issues also exist on XYZEval and in bedlevel.h.
```suggestion
  operator const XYval<T>&()                    const { return *(const XYval<T>*)this; }
```